datasette/app.py

```diff
@@ -588,134 +584,132 @@ class TracingSanic(Sanic):
-            if status != 500:
-                templates = ["{}.html".format(status)] + templates
-            info.update(
-                {"ok": False, "error": message, "status": status, "title": title}
-            )
-            if request is not None and request.path.split("?")[0].endswith(".json"):
-                r = response.json(info, status=status)
-
-            else:
-                template = self.jinja_env.select_template(templates)
-                r = response.html(template.render(info), status=status)
-            if self.cors:
-                r.headers["Access-Control-Allow-Origin"] = "*"
-            return r
+        outer_self = self
+
+        class DatasetteRouter(AsgiRouter):
+            async def handle_404(self, scope, receive, send):
+                # If URL has a trailing slash, redirect to URL without it
+                path = scope.get("raw_path", scope["path"].encode("utf8"))
+                if path.endswith(b"/"):
+                    path = path.rstrip(b"/")
+                    if scope["query_string"]:
+                        path += b"?" + scope["query_string"]
+                    await asgi_send_redirect(send, path.decode("latin1"))
+                else:
+                    await super().handle_404(scope, receive, send)
+
+            async def handle_500(self, scope, receive, send, exception):
+                title = None
+                help = None
+                if isinstance(exception, NotFound):
+                    status = 404
+                    info = {}
+                    message = exception.args[0]
+                elif isinstance(exception, InvalidUsage):
+                    status = 405
+                    info = {}
+                    message = exception.args[0]
+                elif isinstance(exception, DatasetteError):
+                    status = exception.status
+                    info = exception.error_dict
+                    message = exception.message
+                    if exception.messagge_is_html:
+                        message = Markup(message)
+                    title = exception.title
+                else:
+                    status = 500
+                    info = {}
+                    message = str(exception)
+                    traceback.print_exc()
+                templates = ["500.html"]
+                if status != 500:
+                    templates = ["{}.html".format(status)] + templates
+                info.update(
+                    {"ok": False, "error": message, "status": status, "title": title}
+                )
+                headers = {}
+                if outer_self.cors:
+                    headers["Access-Control-Allow-Origin"] = "*"
+                if scope["path"].split("?")[0].endswith(".json"):
+                    await asgi_send_json(send, info, status=status, headers=headers)
+                else:
+                    template = outer_self.jinja_env.select_template(templates)
```

Comment:
The only reason this thing is a nested class is to allow for this access to the `jinja_env` - I could pass this in to the constructor instead.

Reply:
Fixed.